src/Microsoft.ML.Data/Data/RowCursorUtils.cs

```diff
@@ -464,6 +464,30 @@ public static IDataView RowAsDataView(IHostEnvironment env, Row row)
             return new OneRowDataView(env, row);
         }
 
+        /// <summary>
+        /// Given a collection of <see cref="Schema.Column"/>, that is a subset of the Schema of the data, create a predicate,
+        /// that when passed a column index, will return <langword>true</langword> or <langword>false</langword>, based on whether
+        /// the column with the given <see cref="Schema.Column.Index"/> is part of the <paramref name="colsNeeded"/>.
+        /// </summary>
+        /// <param name="colsNeeded">The subset of columns from the <see cref="Schema"/> that are needed from this <see cref="RowCursor"/>.</param>
+        /// <param name="sourceSchema">The <see cref="Schema"/> from where the colsNeeded originate.</param>
+        [BestFriend]
+        internal static Func<int, bool> FromColumnsToPredicate(IEnumerable<Schema.Column> colsNeeded, Schema sourceSchema)
+        {
+            Contracts.CheckValue(colsNeeded, nameof(colsNeeded));
+            Contracts.CheckValue(sourceSchema, nameof(sourceSchema));
+
+            bool[] indicesRequested = new bool[sourceSchema.Count];
+
+            foreach (var col in colsNeeded)
+            {
+                Contracts.Assert(col.Index < indicesRequested.Length, $"The requested column: {col} is not part of the {nameof(sourceSchema)}");
```

Comment:
>Assert [](start = 26, length = 6)

it's a bit strange to have mix of Check and Assert in code

Reply:
Let's review what the three types of things are for:

1. Asserts are basically code documentation. You rarely, to the point of almost never, want string messages associated with them. That's not what they're for.
2. Checks are for things that are actually intended to be runtime compiletime checked -- that is, something that might be checked. But, because C# like nearly all languages does not do lazy evaluation of arguments, I have forbidden passing anything *except* a string literal, since the message is formatted, strings allocated, and everything, whether you actually wind up wanting the error message *or not*.
3. You will note throughout the codebase the use of `if (b) throw Contracts.Except` or something similar in the case. This is primarily for exactly this case where we want to throw a complex message, but we only want to pay the cost of formatting the message in the case where it is actually needed. So this is what you should use here.

So please throw one of the excepts. Assert is wrong since it is for code documentation, documenting invariants, and so on. (Is `ExceptSchemaMismatch` inappropriate here? Maybe it throws the wrong type of message.)

We could be and probably should be a *lot* more aggressive with our checks FYI. Not only should the index be *in range* but the fields or properties of the `Schema.Column` at that index should be identical.

---
In reply to: [247006403](https://github.com/dotnet/machinelearning/pull/1975#discussion_r247006403) [](ancestors = 247006403,246950288)